File: src/main/recovery.ts
/**
 * Zoom Kiosk - Recovery Watchdog
 * 
 * Monitors connection state and automatically attempts to recover
 * after network disconnections or meeting failures.
 */

import { RecoveryConfig } from './config';

/**
 * Recovery state enumeration
 */
enum RecoveryState {
  Idle = 'idle',
  Monitoring = 'monitoring',
  Recovering = 'recovering',
  Failed = 'failed'
}

/**
 * Recovery Watchdog class
 * Implements exponential backoff retry logic for meeting recovery
 */
export class RecoveryWatchdog {
  private config: RecoveryConfig;
  private reconnectCallback: () => Promise<void>;
  private state: RecoveryState = RecoveryState.Idle;
  private retryCount: number = 0;
  private retryTimer: NodeJS.Timeout | null = null;
  private healthCheckTimer: NodeJS.Timeout | null = null;

  constructor(config: RecoveryConfig, reconnectCallback: () => Promise<void>) {
    this.config = config;
    this.reconnectCallback = reconnectCallback;
  }

  /**
   * Start monitoring for disconnections
   */
  start(): void {
    this.state = RecoveryState.Monitoring;
    this.retryCount = 0;
    console.log('[RecoveryWatchdog] Started monitoring');
  }

  /**
   * Stop the watchdog
   */
  stop(): void {
    this.state = RecoveryState.Idle;
    this.clearTimers();
    console.log('[RecoveryWatchdog] Stopped');
  }

  /**
   * Called when a disconnection is detected
   */
  onDisconnected(): void {
    // If already recovering, reset and start fresh recovery
    // This handles cases where a new disconnect happens during recovery
    if (this.state === RecoveryState.Recovering) {
      console.log('[RecoveryWatchdog] New disconnect detected during recovery, resetting and starting fresh');
      this.clearTimers();
      this.retryCount = 0; // Reset retry count for fresh start
    }

    // If in failed state, reset to allow new recovery attempts
    if (this.state === RecoveryState.Failed) {
      console.log('[RecoveryWatchdog] Reset from failed state, starting new recovery');
      this.retryCount = 0;
    }

    console.log('[RecoveryWatchdog] Disconnection detected, starting recovery');
    this.state = RecoveryState.Recovering;
    this.scheduleRetry();
  }

  /**
   * Called when successfully reconnected
   */
  onConnected(): void {
    console.log('[RecoveryWatchdog] Connection restored');
    this.state = RecoveryState.Monitoring;
    this.retryCount = 0;
    this.clearTimers();
  }

  /**
   * Called when sharing is restored
   */
  onSharingRestored(): void {
    console.log('[RecoveryWatchdog] Sharing restored, recovery complete');
    this.state = RecoveryState.Monitoring;
    this.retryCount = 0;
  }

  /**
   * Schedule a retry attempt with exponential backoff
   */
  private scheduleRetry(): void {
    if (this.retryCount >= this.config.maxRetries) {
      console.error('[RecoveryWatchdog] Max retries reached, entering failed state');
      this.state = RecoveryState.Failed;
      return;
    }

    const backoff = this.calculateBackoff();
    console.log(`[RecoveryWatchdog] Scheduling retry ${this.retryCount + 1}/${this.config.maxRetries} in ${backoff}ms`);

    this.retryTimer = setTimeout(async () => {
      await this.attemptRecovery();
    }, backoff);
  }

  /**
   * Calculate exponential backoff delay
   */
  private calculateBackoff(): number {
    const exponentialDelay = this.config.initialBackoffMs * Math.pow(2, this.retryCount);
    const jitter = Math.random() * 1000; // Add random jitter up to 1 second
    return Math.min(exponentialDelay + jitter, this.config.maxBackoffMs);
  }

  /**
   * Attempt to recover the connection
   */
  private async attemptRecovery(): Promise<void> {
    this.retryCount++;
    console.log(`[RecoveryWatchdog] Attempting recovery (attempt ${this.retryCount})`);

    try {
      await this.reconnectCallback();
      // Success - the callback will trigger onConnected via event
    } catch (error) {
      console.error('[RecoveryWatchdog] Recovery attempt failed:', error);
      
      if (this.retryCount < this.config.maxRetries) {
        this.scheduleRetry();
      } else {
        this.state = RecoveryState.Failed;
        console.error('[RecoveryWatchdog] All recovery attempts exhausted');
      }
    }
  }

  /**
   * Clear all timers
   */
  private clearTimers(): void {
    if (this.retryTimer) {
      clearTimeout(this.retryTimer);
      this.retryTimer = null;
    }
    if (this.healthCheckTimer) {
      clearInterval(this.healthCheckTimer);
      this.healthCheckTimer = null;
    }
  }

  /**
   * Get current recovery state
   */
  getState(): RecoveryState {
    return this.state;
  }

  /**
   * Get retry count
   */
  getRetryCount(): number {
    return this.retryCount;
  }

  /**
   * Check if in failed state (needs manual intervention)
   */
  needsAttention(): boolean {
    return this.state === RecoveryState.Failed;
  }

  /**
   * Reset and restart recovery attempts
   */
  reset(): void {
    this.clearTimers();
    this.retryCount = 0;
    this.state = RecoveryState.Monitoring;
    console.log('[RecoveryWatchdog] Reset, ready for new recovery cycle');
  }
}

File: src/recovery.py
"""
Zoom Kiosk - Recovery Watchdog

Monitors connection state and automatically attempts to recover
after network disconnections or meeting failures.
"""

import asyncio
import random
from typing import Callable, Awaitable
from .config import RecoveryConfig


class RecoveryState:
    IDLE = 'idle'
    MONITORING = 'monitoring'
    RECOVERING = 'recovering'
    FAILED = 'failed'


class RecoveryWatchdog:
    """Recovery Watchdog class with exponential backoff retry logic"""

    def __init__(self, config: RecoveryConfig, reconnect_callback: Callable[[], Awaitable[None]]):
        self.config = config
        self.reconnect_callback = reconnect_callback
        self.state = RecoveryState.IDLE
        self.retry_count = 0
        self.retry_task: asyncio.Task | None = None

    def start(self) -> None:
        """Start monitoring for disconnections"""
        self.state = RecoveryState.MONITORING
        self.retry_count = 0
        print('[RecoveryWatchdog] Started monitoring')

    def stop(self) -> None:
        """Stop the watchdog"""
        self.state = RecoveryState.IDLE
        self._clear_timers()
        print('[RecoveryWatchdog] Stopped')

    def on_disconnected(self) -> None:
        """Called when a disconnection is detected"""
        # If already recovering, reset and start fresh recovery
        if self.state == RecoveryState.RECOVERING:
            print('[RecoveryWatchdog] New disconnect detected during recovery, resetting and starting fresh')
            self._clear_timers()
            self.retry_count = 0

        # If in failed state, reset to allow new recovery attempts
        if self.state == RecoveryState.FAILED:
            print('[RecoveryWatchdog] Reset from failed state, starting new recovery')
            self.retry_count = 0

        print('[RecoveryWatchdog] Disconnection detected, starting recovery')
        self.state = RecoveryState.RECOVERING
        self._schedule_retry()

    def on_connected(self) -> None:
        """Called when successfully reconnected"""
        print('[RecoveryWatchdog] Connection restored')
        self.state = RecoveryState.MONITORING
        self.retry_count = 0
        self._clear_timers()

    def on_sharing_restored(self) -> None:
        """Called when sharing is restored"""
        print('[RecoveryWatchdog] Sharing restored, recovery complete')
        self.state = RecoveryState.MONITORING
        self.retry_count = 0

    def _schedule_retry(self) -> None:
        """Schedule a retry attempt with exponential backoff"""
        if self.retry_count >= self.config["maxRetries"]:
            print('[RecoveryWatchdog] Max retries reached, entering failed state')
            self.state = RecoveryState.FAILED
            return

        backoff = self._calculate_backoff()
        print(f'[RecoveryWatchdog] Scheduling retry {self.retry_count + 1}/{self.config["maxRetries"]} in {backoff}ms')

        async def retry_task():
            await asyncio.sleep(backoff / 1000.0)
            await self._attempt_recovery()

        self.retry_task = asyncio.create_task(retry_task())

    def _calculate_backoff(self) -> float:
        """Calculate exponential backoff delay"""
        exponential_delay = self.config["initialBackoffMs"] * (2 ** self.retry_count)
        jitter = random.random() * 1000  # Add random jitter up to 1 second
        return min(exponential_delay + jitter, self.config["maxBackoffMs"])

    async def _attempt_recovery(self) -> None:
        """Attempt to recover the connection"""
        self.retry_count += 1
        print(f'[RecoveryWatchdog] Attempting recovery (attempt {self.retry_count})')

        try:
            await self.reconnect_callback()
            # Success - the callback will trigger on_connected via event
        except Exception as e:
            print(f'[RecoveryWatchdog] Recovery attempt failed: {e}')

            if self.retry_count < self.config["maxRetries"]:
                self._schedule_retry()
            else:
                self.state = RecoveryState.FAILED
                print('[RecoveryWatchdog] All recovery attempts exhausted')

    def _clear_timers(self) -> None:
        """Clear all timers"""
        if self.retry_task and not self.retry_task.done():
            self.retry_task.cancel()
            self.retry_task = None

    def get_state(self) -> str:
        """Get current recovery state"""
        return self.state

    def get_retry_count(self) -> int:
        """Get retry count"""
        return self.retry_count

    def needs_attention(self) -> bool:
        """Check if in failed state (needs manual intervention)"""
        return self.state == RecoveryState.FAILED

    def reset(self) -> None:
        """Reset and restart recovery attempts"""
        self._clear_timers()
        self.retry_count = 0
        self.state = RecoveryState.MONITORING
        print('[RecoveryWatchdog] Reset, ready for new recovery cycle')
